File: src/main/org/nlogo/nvm/Argument.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.nvm;

import org.nlogo.agent.AgentSet;
import org.nlogo.api.Dump;
import org.nlogo.api.Syntax;
import org.nlogo.api.TypeNames;
import java.util.List;

/**
 * Passes arguments to extension primitives.
 */
public strictfp class Argument
    implements org.nlogo.api.Argument {
  private final Context context;
  protected final Reporter arg;
  private Object cached = null;

  public Argument(Context context, Reporter arg) {
    this.context = context;
    this.arg = arg;
  }

  public Object get()
      throws org.nlogo.api.LogoException {
    if (cached == null) {
      cached = arg.report(context);
    }
    if (cached instanceof org.nlogo.agent.Agent &&
        ((org.nlogo.agent.Agent) cached).id == -1) {
      cached = org.nlogo.api.Nobody$.MODULE$;
    }
    return cached;
  }

  public AgentSet getAgentSet()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (AgentSet) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException
          (getExceptionMessage
           (Syntax.AgentsetType(), obj));
    }
  }


  public org.nlogo.api.Agent getAgent()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Agent) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.AgentType(), obj));
    }
  }

  public Boolean getBoolean()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (Boolean) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.BooleanType(), obj));
    }
  }

  public boolean getBooleanValue()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return ((Boolean) obj).booleanValue();
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.BooleanType(), obj));
    }
  }

  public double getDoubleValue()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return ((Double) obj).doubleValue();
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.NumberType(), obj));
    }
  }

  public int getIntValue()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return ((Double) obj).intValue();
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.NumberType(), obj));
    }
  }

  public org.nlogo.api.LogoList getList()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.LogoList) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.ListType(), obj));
    }
  }

  public org.nlogo.api.Patch getPatch()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Patch) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.PatchType(), obj));
    }
  }

  public String getString()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (String) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.StringType(), obj));
    }
  }


  public org.nlogo.api.Turtle getTurtle()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Turtle) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.TurtleType(), obj));
    }
  }

  public org.nlogo.api.Link getLink()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.Link) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.LinkType(), obj));
    }
  }

  public org.nlogo.api.ReporterTask getReporterTask()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.ReporterTask) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.ReporterTaskType(), obj));
    }
  }

  public org.nlogo.api.CommandTask getCommandTask()
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (org.nlogo.api.CommandTask) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.CommandTaskType(), obj));
    }
  }

  @SuppressWarnings("unchecked") public List<org.nlogo.api.Token> getCode() 
      throws org.nlogo.api.ExtensionException, org.nlogo.api.LogoException {
    Object obj = get();
    try {
      return (List<org.nlogo.api.Token>) obj;
    } catch (ClassCastException ex) {
      throw new org.nlogo.api.ExtensionException(
        getExceptionMessage(Syntax.CodeBlockType(), obj));
    }
  }

  /**
   * <i>Special (undocumented) method for the cities extension. </i>
   * Returns the argument reporter without evaluating.
   */
  public Reporter getReporter() {
    return arg;
  }


  private String getExceptionMessage(int wantedType, Object badValue) {
    String result = "Expected this input to be "
        + TypeNames.aName(wantedType) + " but got "
        + (badValue == org.nlogo.api.Nobody$.MODULE$
        ? "NOBODY"
        : "the " + TypeNames.name(badValue)
        + " " + Dump.logoObject(badValue))
        + " instead.";

    return result;

  }

}

File: src/main/org/nlogo/nvm/Argument.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.nvm

import org.nlogo.api
import api.{ Dump, Nobody, Syntax, TypeNames, LogoException, ExtensionException }

/**
 * Passes arguments to extension primitives.
 */
class Argument(context: Context, arg: Reporter) extends api.Argument {

  private[this] var cached: AnyRef = null

  @throws(classOf[LogoException])
  def get = {
    if (cached == null)
      cached = arg.report(context)
    cached match {
      case a: api.Agent if a.id == -1 =>
        cached = Nobody
      case _ =>
    }
    cached
  }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getAgentSet: api.AgentSet =
    get match {
      case agents: api.AgentSet => agents
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.AgentsetType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getAgent: api.Agent =
    get match {
      case agent: api.Agent => agent
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.AgentType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getBoolean: java.lang.Boolean =
    get match {
      case b: java.lang.Boolean => b
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.BooleanType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getBooleanValue: Boolean =
    get match {
      case b: java.lang.Boolean => b.booleanValue
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.BooleanType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getDoubleValue: Double =
    get match {
      case d: java.lang.Double => d.doubleValue
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.NumberType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getIntValue: Int =
    get match {
      case d: java.lang.Double => d.intValue
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.NumberType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getList: api.LogoList =
    get match {
      case l: api.LogoList => l
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.ListType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getPatch: api.Patch =
    get match {
      case p: api.Patch => p
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.PatchType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getString: String =
    get match {
      case s: String => s
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.StringType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getTurtle: api.Turtle =
    get match {
      case t: api.Turtle => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.TurtleType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getLink: api.Link =
    get match {
      case l: api.Link => l
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.LinkType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getReporterTask: api.ReporterTask =
    get match {
      case t: api.ReporterTask => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.ReporterTaskType, x))
    }

  @throws(classOf[ExtensionException])
  @throws(classOf[LogoException])
  def getCommandTask: api.CommandTask =
    get match {
      case t: api.CommandTask => t
      case x =>
        throw new ExtensionException(
          getExceptionMessage(Syntax.CommandTaskType, x))
    }

  /**
   * <i>Special (undocumented) method for the cities extension.</i>
   * Returns the argument reporter without evaluating it.
   */
  def getReporter: Reporter = arg

  def getExceptionMessage(wantedType: Int, badValue: AnyRef) =
    "Expected this input to be " +
      TypeNames.aName(wantedType) + " but got " +
       (if(badValue == api.Nobody) "NOBODY"
        else ("the " + TypeNames.name(badValue) + " " +
              Dump.logoObject(badValue)) +
              " instead.")

}
